tests/phpunit/Civi/Angular/LoaderTest.php

```diff
@@ -0,0 +1,97 @@
+<?php
+/*
+ +--------------------------------------------------------------------+
+ | Copyright CiviCRM LLC. All rights reserved.                        |
+ |                                                                    |
+ | This work is published under the GNU AGPLv3 license with some      |
+ | permitted exceptions and without any warranty. For full license    |
+ | and copyright information, see https://civicrm.org/licensing       |
+ +--------------------------------------------------------------------+
+ */
+
+namespace Civi\Angular;
+
+/**
+ * Test the Angular loader.
+ */
+class LoaderTest extends \CiviUnitTestCase {
+
+  public static $dummy_setting_count = 0;
+  public static $dummy_callback_count = 0;
+
+  public function setUp() {
+    parent::setUp();
+    $this->hookClass->setHook('civicrm_angularModules', [$this, 'hook_angularModules']);
+    self::$dummy_setting_count = 0;
+    self::$dummy_callback_count = 0;
+    $this->createLoggedInUser();
+  }
+
+  public function factoryScenarios() {
+    return [
+      ['dummy1', 2, 1],
+      ['dummy2', 2, 0],
+      ['dummy3', 2, 2],
+    ];
+  }
+
+  /**
+   * Tests that AngularLoader only conditionally loads settings via factory functions for in-use modules.
+   * Our dummy settings callback functions keep a count of the number of times they have been called.
+   *
+   * @dataProvider factoryScenarios
+   * @param $module
+   * @param $expectedSettingCount
+   * @param $expectedCallbackCount
+   */
+  public function testSettingFactory($module, $expectedSettingCount, $expectedCallbackCount) {
+    $loader = new \Civi\Angular\AngularLoader();
+    $loader->setModules([$module]);
+    $loader->useApp();
+    $loader->load();
+    \Civi::resources()->getSettings();
+
+    $this->assertEquals($expectedSettingCount, self::$dummy_setting_count);
+    $this->assertEquals($expectedCallbackCount, self::$dummy_callback_count);
```

Comment:
I like this test a lot, especially how it covers the various mixes of using/not-using options..

One thing - the assert here is a little indirect, e.g. checking that the function was called. Could it also assert about the intended outcome? To wit: `getSettings()['foo']` should be defined if and only if `$expectFooCount` is greater than 0.

```php
$all = Civi::resources()->getSettings();

$this->assertTrue($expectedSettingCount > 0 === isset($all['dummy_setting']));
$this->assertTrue($expectedCallbackCount > 0 === isset($all['dummy_setting_factory']));
```